pulsar-client/src/main/java/org/apache/pulsar/client/impl/transaction/TransactionImpl.java

```diff
@@ -117,7 +117,7 @@ public synchronized void registerSendOp(CompletableFuture<MessageId> sendFuture)
         return checkIfOpen().thenCompose(value -> {
             synchronized (TransactionImpl.this) {
                 // we need to issue the request to TC to register the acked topic
-                return registerSubscriptionMap.compute(topic, (key, future) -> {
+                return registerSubscriptionMap.compute(topic + "-" + subscription, (key, future) -> {
```

Comment:
This will bring risks when using `-` to connect the topic name and subscription name. For example, if we have a topic `a-b` and subscription name `c`, the key will be `a-b-c`, for topic `a` and subscription `b-c`, we also get the key `a-b-c`.

You can use Pair.of(topic, sub) as the key, or use a special connector which is not allowed in Pulsar topic name and sub name.